server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -593,11 +594,22 @@ private void cleanupAfterError(Exception exception) {
         }
     }
 
-    private static ShardGenerations buildGenerations(SnapshotsInProgress.Entry snapshot) {
+    private static ShardGenerations buildGenerations(SnapshotsInProgress.Entry snapshot, MetaData metaData) {
         ShardGenerations.Builder builder = ShardGenerations.builder();
         final Map<String, IndexId> indexLookup = new HashMap<>();
-        snapshot.indices().forEach(idx -> indexLookup.put(idx.getName(), idx));
-        snapshot.shards().forEach(c -> builder.put(indexLookup.get(c.key.getIndexName()), c.key.id(), c.value.generation()));
+        snapshot.indices().forEach(idx -> {
+            if (metaData.index(idx.getName()) != null) {
```

Comment:
It's a bit annoying that we have to lookup by index name here, not index uuid. In theory, an index could have been deleted and recreated between snapshot start and completion. Do we have index uuid perhaps in the shard snapshots?

Reply:
Good point, this is actually a possibility (I reproduced it with the test added in d219dfa6213) and fixed it now by using the `Index` instances in the shards map from `SnapshotInProgress` :)